File: lib/csharp/microsoft.bot.builder.solutions/microsoft.bot.builder.solutions/Responses/SpeechUtility.cs
using System;
using System.Collections.Generic;
using AdaptiveCards;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Solutions.Extensions;
using Microsoft.Bot.Builder.Solutions.Resources;
using Microsoft.Bot.Schema;

namespace Microsoft.Bot.Builder.Solutions.Responses
{
    /// <summary>
    /// Read order of list items.
    /// </summary>
    public enum ReadPreference
    {
        /// <summary>First item, second item, third item, etc.</summary>
        Enumeration,

        /// <summary>Latest item, second item, third item, etc.</summary>
        Chronological
    }

    public class SpeechUtility
    {
        /// <summary>
        /// Concatenate PromptOption string properties into a formatted speech-ready string.
        /// </summary>
        /// <param name="selectOption">Prompt options.</param>
        /// <param name="readOrder">Read order of list items.</param>
        /// <param name="maxSize">The max read size, default is 4.</param>
        /// <returns>Formatted speech-ready string.</returns>
        public static string ListToSpeechReadyString(PromptOptions selectOption, ReadPreference readOrder = ReadPreference.Enumeration,  int maxSize = 4)
        {
            List<string> selectOptionSpeakStrings = new List<string>();

            for (int i = 0; i < selectOption.Choices.Count; ++i)
            {
                selectOptionSpeakStrings.Add(selectOption.Choices[i].Value);
            }

            return ListToSpeechReadyString(selectOption.Prompt.Text, selectOptionSpeakStrings, readOrder, maxSize);
        }

        /// <summary>
        /// Concatenate Activity string properties into a formatted speech-ready string.
        /// </summary>
        /// <param name="activityToProcess">Activity.</param>
        /// <param name="readOrder">Read order of list items.</param>
        /// <param name="maxSize">The max read size, default is 4.</param>
        /// <returns>Formatted speech-ready string.</returns>
        public static string ListToSpeechReadyString(Activity activityToProcess, ReadPreference readOrder = ReadPreference.Enumeration, int maxSize = 4)
        {
            List<string> selectOptionSpeakStrings = new List<string>();

            for (int i = 0; i < activityToProcess.Attachments.Count; ++i)
            {
                // Add try parse - if already an adaptive card good to go, if unknown object convert from string
                var cardContent = AdaptiveCard.FromJson(activityToProcess.Attachments[i].Content.ToString());
                if (!string.IsNullOrEmpty(cardContent?.Card?.Speak))
                {
                    selectOptionSpeakStrings.Add(cardContent.Card.Speak);
                }
            }

            return ListToSpeechReadyString(activityToProcess.Speak, selectOptionSpeakStrings, readOrder, maxSize);
        }

        /// <summary>
        /// Concatenate strings into a formatted speech-ready string.
        /// </summary>
        /// <param name="parentString">Itroduction string.</param>
        /// <param name="selectionStrings">List item strings.</param>
        /// <param name="readOrder">Read order of list items.</param>
        /// <param name="maxSize">The max read size.</param>
        /// <returns>Formatted speech-ready string.</returns>
        private static string ListToSpeechReadyString(string parentString, List<string> selectionStrings, ReadPreference readOrder, int maxSize)
        {
            var result = string.Empty;
            if (!string.IsNullOrEmpty(parentString))
            {
                result += parentString + "<break/>";
            }

            List<string> itemDetails = new List<string>();

            int readSize = Math.Min(selectionStrings.Count, maxSize);
            if (readSize == 1)
            {
                itemDetails.Add(selectionStrings[0]);
            }
            else
            {
                for (var i = 0; i < readSize; i++)
                {
                    var readFormat = string.Empty;

                    if (i == 0)
                    {
                        if (readOrder.Equals(ReadPreference.Chronological))
                        {
                            readFormat = CommonStrings.LatestItem;
                        }
                        else
                        {
                            readFormat = CommonStrings.FirstItem;
                        }
                    } else
                    {
                        if (i == readSize - 1)
                        {
                            readFormat = CommonStrings.LastItem;
                        }
                        else
                        {
                            if (i == 1)
                            {
                                readFormat = CommonStrings.SecondItem;
                            }
                            else if (i == 2)
                            {
                                readFormat = CommonStrings.ThirdItem;
                            }
                            else if (i == 3)
                            {
                                readFormat = CommonStrings.FourthItem;
                            }
                        }
                    }

                    var selectionDetail = string.Format(readFormat, selectionStrings[i]);
                    itemDetails.Add(selectionDetail);
                }
            }

            result += itemDetails.ToSpeechString(CommonStrings.And);
            return result;
        }
    }
}

File: sdk/typescript/libraries/bot-solutions/src/responses/speechUtility.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { Choice, PromptOptions } from 'botbuilder-dialogs';
import { Activity, Attachment } from 'botframework-schema';
import { ListEx } from '../extensions';
import { CommonResponses } from '../resources';
import { readFileSync } from 'fs';
import { ResponsesUtil } from '../util/responsesUtil';

/**
 * Read order of list items.
 */
export enum ReadPreference {
    /**
     * First item, second item, third item, etc.
     */
    Enumeration,
    /**
     * Latest item, second item, third item, etc.
     */
    Chronological
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace SpeechUtility {

    export function listToSpeechReadyString(
        toProcess: PromptOptions|Activity,
        locale: string,
        readOrder: ReadPreference = ReadPreference.Enumeration,
        maxSize: number = 4
    ): string {

        let speakStrings: string[] = [];
        let parent = '';

        if ((toProcess as PromptOptions).choices !== undefined) {
            const selectOption: PromptOptions = toProcess as PromptOptions;
            const choices: (string|Choice)[] = selectOption.choices || [];
            speakStrings = choices.map((value: string|Choice): string => typeof(value) === 'string' ? value : value.value);
            parent = typeof(selectOption.prompt) === 'string'
                ? (selectOption.prompt || '')
                : (selectOption.prompt ? (selectOption.prompt.text || '') : '');
        } else {
            const activity: Activity = toProcess as Activity;
            const attachments: Attachment[] = activity.attachments || [];
            speakStrings = attachments.map((value: Attachment): string => value.content.speak as string);
            parent = activity.speak || '';
        }

        return listToSpeech(parent, speakStrings, readOrder, maxSize, locale);
    }

    function listToSpeech(parent: string, selectionStrings: string[], readOrder: ReadPreference, maxSize: number, locale: string): string {

        const result: string = `${ parent } ` || '';

        const itemDetails: string[] = [];
        const readSize: number = Math.min(selectionStrings.length, maxSize);
        const jsonPath: string = ResponsesUtil.getResourcePath(CommonResponses.name, CommonResponses.pathToResource, locale);
        const commonFile: string = readFileSync(jsonPath, 'utf8');

        if (readSize === 1) {
            itemDetails.push(selectionStrings[0]);
        } else {
            for (let index = 0; index < readSize; index = index + 1) {
                let readFormat = '';
                if (index === 0) {
                    if (readOrder === ReadPreference.Chronological) {
                        readFormat = JSON.parse(commonFile)['latestItem'];
                    } else {
                        readFormat = JSON.parse(commonFile)['firstItem'];
                    }
                } else {
                    if (index === readSize - 1) {
                        readFormat = JSON.parse(commonFile)['lastItem'];
                    } else {
                        if (index === 1) {
                            readFormat = JSON.parse(commonFile)['secondItem'];
                        } else if (index === 2) {
                            readFormat = JSON.parse(commonFile)['thirdItem'];
                        } else if (index === 3) {
                            readFormat = JSON.parse(commonFile)['fourthItem'];
                        }
                    }
                }

                const selectionDetail: string = readFormat.replace('{0}', selectionStrings[index]);
                itemDetails.push(selectionDetail);
            }
        }


        return result + ListEx.toSpeechString(itemDetails, JSON.parse(commonFile)['and'], locale);
    }
}
